sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc

```diff
@@ -0,0 +1,293 @@
+= sycl_ext_oneapi_group_load_store
+
+:source-highlighter: coderay
+:coderay-linenums-mode: table
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+:dpcpp: pass:[DPC++]
+:language: {basebackend@docbook:c++:cpp}
+
+== Introduction
+
+IMPORTANT: This specification is a draft.
+
+
+== Notice
+
+[%hardbreaks]
+Copyright (c) 2022 Intel Corporation.  All rights reserved.
+
+Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
+of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
+permission by Khronos.
+
+== Dependencies
+
+This extension is written against the SYCL 2020 specification, Revision 6.
+All references below to the "core SYCL specification" or to section numbers in
+the SYCL specification refer to that revision.
+
+This extension also depends on the following other SYCL extensions:
+
+* link:https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_group_sort.asciidoc[
+  sycl_ext_oneapi_group_sort] (`group_with_scratchpad` class used as GroupHelper).
+
+* link:../experimental/sycl_ext_oneapi_properties.asciidoc[
+  sycl_ext_oneapi_properties].
+
+== Overview
+
+This extension defines free functions for load/store operations within work
+group scope.
+
+== Single Value Functions
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+// Load API scalar
+template<typename Group, typename InputIteratorT,
+         typename OutputT, typename Properties = properties<>>
+void group_load(Group g, InputIteratorT in_ptr,
+                      OutputT& out, Properties = {}); (1)
+
+// Store API scalar
+template<typename Group, typename InputT,
+         typename OutputIteratorT, typename Properties = properties<>>
+void group_store(Group g, const InputT& in,
+                      OutputIteratorT out_ptr, Properties = {}); (2)
+
+}
+----
+
+(1) _Preconditions_: `in_ptr` must be the same for all work-items
+in the group.
+
+_Constraints_: Only available if `Group` is a work-group or sub-group.
+
+_Effects_: Loads single element from `in_ptr` to `out`.
```

Comment:
Add "using the g group object" or something like that because we don't specify how to use `g` here.

Reply:
Good point! Actually, I noticed an inconsistency in my understanding here with [the previous proposal](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_group_load_store.asciidoc), @andreyfe1, @Pennycook, please, advice:

I thought that for single value API the following test-case should work ok: https://github.com/intel/llvm-test-suite/blob/8950ae921bd2be785d42de113a278f91571d830e/SYCL/GroupLoadStore/group_load_store_api.cpp#L19
and we don't need to pass different pointers for different groups and use addresses like: `ptr + g.get_group_id(0) * g.get_local_range(0)  + g.get_local_id(0)`.

But if it should work as `ptr + get_local_id()`, then if we want to load from global memory for each work item, it will look like this:

```
      cgh.parallel_for(sycl::nd_range<1>(size, block_size),
                       [=](sycl::nd_item<1> item) {
                         auto g = item.get_group();

                         InputT data;

                         sycl_exp::group_load(g, in.get_pointer() +  g.get_group_id(0) * g.get_local_range(0), data);

                         data += item.get_global_linear_id() * 100000;

                         sycl_exp::group_store(g, data, out.get_pointer() + g.get_group_id(0) * g.get_local_range(0));
                       });
    });
```
or is it just not the primary use case we want to cover? As we assume the same behaviour as mentioned above for sycl::vec and span-based API:

```
      cgh.parallel_for(
          sycl::nd_range<1>(block_count * block_size, block_size),
          [=](sycl::nd_item<1> item) {
            InputT data[items_per_thread];

            sycl_exp::group_load(item.get_group(), in.get_pointer(),
                                 sycl::span{data});

            for (int i = 0; i < items_per_thread; ++i) {
              data[i] += item.get_global_linear_id() * 100000 + i * 1000;
            }

            sycl_exp::group_store(item.get_group(), sycl::span{data},
                                  out.get_pointer());
          });
```
It will load a contiguous block of memory from the single global pointer by default:
|         wi[0]         |             wi[1]    |          wi[2]        |
|-----------------|------------------|-----------------|
| in[0] in[1] in[2] | in[3] in[4] in[5] | in[6] in[7] in[8] |

in striped case:
|         wi[0]         |             wi[1]    |          wi[2]        |
|-----------------|------------------|-----------------|
| in[0] in[3] in[6] | in[1] in[4] in[7] | in[2] in[5] in[8] |

Reply:
or it is still a matter of naming and we should distinguish group_load/store, which works with local_id() in the group and block_load/store which works with global_id()

Reply:
I think it's important that the group algorithms assume that only the group passed into the function is involved in the call.  If the behavior was defined such that `sycl::group_load(g, ptr)` looked at the _global_ work-item ID, then the only way to access all the data in `ptr` would be to ensure that all the groups execute the `load` function.  I think most developers write code from the perspective of a single work-item, and so they're used to having to calculate things like per-group offsets.

The use-case that you mentioned is important, though, and the planned solution is to provide different types of group (e.g. see [sycl_ext_oneapi_root_group](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_root_group.asciidoc)):

```c++
sycl::root_group = it.get_root_group();
sycl::group_load(root_group, in.get_pointer(), sycl::span{data});
```

The `root_group` is a group containing all of the work-items in the kernel, and so calling `root_group::local_id` is equivalent to calling `group::global_id`.

Reply:
ooh, I see, that is a cool proposal. I will align with the initial one, add a detailed description for the expected behaviour and update the tests, thank you!

Reply:
both the proposal and test cases are updated to work with the local scope